Autofill "Never Save for this Site" feature

amddg44 opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/0/1205783642285427/f
Tech Design URL: https://app.asana.com/0/481882893211075/1205590957854802/f
CC:

**Description**:
Adds the functionality to allow users to choose to "Never Save for this Site" from the autofill save login prompt, with the option to reset this list from the Logins screen.

<!--
Tagging instructions
If this PR isn't ready to be merged for whatever reason it should be marked with the `DO NOT MERGE` label (particularly if it's a draft)
If it's pending Product Review/PFR, please add the `Pending Product Review` label.

If at any point it isn't actively being worked on/ready for review/otherwise moving forward (besides the above PR/PFR exception) strongly consider closing it (or not opening it in the first place). If you decide not to close it, make sure it's labelled to make it clear the PRs state and comment with more information.
-->

**Steps to test this PR**:
Note - I'm still waiting on translations so that will probably be a separate PR

1. Visit https://fill.dev/form/login-simple and login using new credentials. In the `Save Login` prompt, confirm the bottom button is now `Never Ask for This Site` 
<img width='200' src="https://github.com/duckduckgo/iOS/assets/91189922/8d991d91-ddda-423c-ac45-3882a3b2b3fc"/>

2. Tap it and confirm this pixel is fired:` m_autofill_logins_save_login_exclude_site_confirmed` 
3. Go back to https://fill.dev/form/login-simple
4. Login again with new credentials and confirm that you no longer get the `Save Login` prompt
5. Visit another site and, logging with new credentials, confirm the `Save Login` prompt is still presented for other sites
6. Go to https://fill.dev/form/registration-email and confirm that `Password Generation` is disabled for this site (no key in the password field and tapping in the field does not trigger the generated password prompt) -- this is why we needed to adopt support for `getRuntimeConfiguration` now
7. Visit another website with a registration form and confirm that `Password Generation` is still enabled for it 
8. Go the Logins screen and you should see a new row with the option to `Reset Excluded Sites` <img width='400' src="https://github.com/duckduckgo/iOS/assets/91189922/43bc5dd6-106a-46cc-994b-6434327645ce"/>

9. Tap the row and confirm this pixel is fired: `m_autofill_settings_reset_excluded_displayed`
10. Tap `Cancel` in the action sheet that was presented and confirm this pixel is fired `m_autofill_settings_reset_excluded_dismissed`
11. Tap the row and choose `Reset Excluded Sites` and confirm this pixel is fired: `m_autofill_settings_reset_excluded_confirmed`
12. The `Reset Excluded Sites` row should now be gone (it is only added to the screen when a user has at least 1 site saved as `Never Prompt`)
13. Go back to https://fill.dev/form/login-simple and confirm the Save Login prompt now appears again 
14. Final test case (a small bug fix): confirm Private Duck Addresses are not autosaved when Autofill is disabled 🙂 

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

amddg44 committed:
Adds Never Prompt option to the Save Login prompt

amddg44 committed:
New Never prompt related strings

amddg44 committed:
Adds dynamically displayed reset never prompt option to AutofillLogSettingsListViewController

amddg44 committed:
Control whether or not save prompt should be presented based on never prompt status

amddg44 committed:
Set BSK branch

amddg44 committed:
Dynamically enabling / disabling the generated password key based on current webview never prompt status

amddg44 committed:
Undoing this attempt at dyanmic password key presentation in favour of new approach

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/autofill-never-prompt-save

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/autofill-never-prompt-save

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Ship review copy updates

amddg44 committed:
Reset excluded sites & related pixels

amddg44 committed:
Layout updated following ship review change requests

amddg44 committed:
AutofillNeverPromptWebsitesManager created + implemented

amddg44 committed:
Pointing to BSK branch

amddg44 committed:
Updates to support getRuntimeConfiguration

amddg44 committed:
Fix for bug where private duck addresses are autosaving when Autofill is disabled in settings

amddg44 committed:
MockSecureVault updated for NeverPromptWebsites

amddg44 committed:
Fix linter warnings

amddg44 committed:
AutofillNeverPromptWebsitesManager created to manage neverPromptWebsites queries

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/autofill-never-prompt-save

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Tests added for AutofillNeverPromptWebsitesManager + AutofillLoginListViewModel

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/autofill-never-prompt-save

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Update BSK revision

amddg44 committed:
Fix failing unit test

afterxleep commented on DuckDuckGo/AutofillLoginSettingsListViewController.swift:
```diff
@@ -470,6 +487,13 @@ final class AutofillLoginSettingsListViewController: UIViewController {
     }
+
+    private func neverSavedCell(for tableView: UITableView, indexPath: IndexPath) -> AutofillNeverSavedTableViewCell {
+        // TODO - update these cells for DRK styling
```
Are you planning to do this as part of this change?

afterxleep commented on DuckDuckGo/AutofillNeverSavedTableViewCell.swift:
```diff
@@ -0,0 +1,80 @@
+        let label = UILabel(frame: CGRect.zero)
+        label.font = .preferredFont(forTextStyle: .callout)
+        label.textColor = .gray90
+//        label.textColor = UIColor(designSystemColor: .label)
```
Assuming this is related to the DRK required changes, but maybe remove for now?

afterxleep reviewed (approved):
Looking good 🚀.   Just a couple nits about commented stuff

amddg44 committed:
Minor styling update

amddg44 committed:
Translations for Autofill "Never Save for this Site" feature (#2157)

Task/Issue URL: https://app.asana.com/0/0/1205159380096691/f
Tech Design URL:
CC:

Description:
Translations for "Never save for this Site" autofill functionality

amddg44 committed:
Autofill prompt icon + copy updates (#2167)

Task/Issue URL: https://app.asana.com/0/72649045549333/1205786958985456/f
Tech Design URL:
CC:

Description:
Removes the website favicon + url at the top of the autofill prompts and add app icon to the top of all autofill prompts. Also updates copy to replace the word logins with passwords

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/autofill-never-prompt-save

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Bump BSK revision

amddg44 committed:
Set to BSK release 84.1.0
